diesel/src/migrations/migration.rs

```diff
@@ -0,0 +1,166 @@
+use connection::Connection;
+use super::{MigrationError, RunMigrationsError};
+
+use std::path::{Path, PathBuf};
+
+pub trait Migration {
+    fn version(&self) -> i64;
+    fn run(&self, conn: &Connection) -> Result<(), RunMigrationsError>;
+    fn revert(&self, conn: &Connection) -> Result<(), RunMigrationsError>;
+}
+
+pub fn migration_from(path: PathBuf) -> Result<Box<Migration>, MigrationError> {
+    if valid_sql_migration_directory(&path) {
+        let version = try!(version_from_path(&path));
+        Ok(Box::new(SqlFileMigration(path, version)))
+    } else {
+        Err(MigrationError::UnknownMigrationFormat(path))
+    }
+}
+
+fn valid_sql_migration_directory(path: &Path) -> bool {
+    macro_rules! t { ($e:expr) => {
+        match $e {
+            Ok(e) => e, Err(_) => return false,
+        }
+    } }
+
+    t!(path.read_dir()).all(|e| {
+        let entry = t!(e);
+        let file_name = entry.file_name();
+        &file_name == "up.sql" || &file_name == "down.sql"
+    }) && t!(path.read_dir()).count() == 2
```

Comment:
Should we potentially make this `>= 2`? In case some projects or companies want to be able to have other files in the dir, like `example_output.txt`?